mypy/plugins/dataclasses.py

```diff
@@ -0,0 +1,279 @@
+            order_other_type = TypeVarType(order_tvar_def)
+            order_return_type = ctx.api.named_type('__builtins__.bool')
+            order_args = [
+                Argument(Var('other', order_other_type), order_other_type, None, ARG_POS)
+            ]
+
+            for method_name in ['__lt__', '__gt__', '__le__', '__ge__']:
+                existing_method = info.get(method_name)
+                if existing_method is not None:
+                    assert existing_method.node
+                    ctx.api.fail(
+                        'You may not have a custom %s method when order=True' % method_name,
+                        existing_method.node,
+                    )
+
+                _add_method(
+                    ctx,
+                    method_name,
+                    args=order_args,
+                    return_type=order_return_type,
+                    self_type=order_other_type,
+                    tvar_def=order_tvar_def,
+                )
+
+        if decorator_arguments['frozen']:
+            self._freeze(attributes)
+
+        info.metadata['dataclass'] = {
+            'attributes': OrderedDict((attr.name, attr.serialize()) for attr in attributes),
+            'frozen': decorator_arguments['frozen'],
+        }
+
+    def collect_attributes(self) -> List[DataclassAttribute]:
+        """Collect all attributes declared in the dataclass and its parents.
+
+        All assignments of the form
+
+          a: SomeType
+          b: SomeOtherType = ...
+
+        are collected.
+        """
+        # First, collect attributes belonging to the current class.
+        ctx = self._ctx
+        cls = self._ctx.cls
+        attrs = []  # type: List[DataclassAttribute]
+        known_attrs = set()  # type: Set[str]
+        for stmt in cls.defs.body:
+            # Any assignment that doesn't use the new type declaration
+            # syntax can be ignored out of hand.
+            if not (isinstance(stmt, AssignmentStmt) and stmt.new_syntax):
+                continue
+
+            # a: int, b: str = 1, 'foo' is not supported syntax so we
+            # don't have to worry about it.
+            lhs = stmt.lvalues[0]
+            if not isinstance(lhs, NameExpr):
+                continue
+
+            try:
+                node = cls.info.names[lhs.name].node
+                assert isinstance(node, Var)
+
+                # x: ClassVar[int] is ignored by dataclasses.
+                if node.is_classvar:
+                    continue
+
+                # Treat the assignment as an instance-level assignment.
+                node.is_initialized_in_class = False
+            except KeyError:
+                continue
+
+            has_field_call, field_args = _collect_field_args(stmt.rvalue)
+
+            try:
+                is_in_init = bool(ctx.api.parse_bool(field_args['init']))
+            except KeyError:
```

Comment:
There are too many things inside this except block. Move all the unnecessary code outside. Also I would probably use `if 'init' in field_args` here.